packages/channel-provider/src/channel-provider.ts

```diff
@@ -49,48 +56,42 @@ class ChannelProvider implements ChannelProviderInterface {
     return response;
   }
 
-  async subscribe(subscriptionType: string): Promise<string> {
+  async subscribe(subscriptionType: Notification['method']): Promise<string> {
     const subscriptionId = Guid.create().toString();
-    if (!this.subscriptions[subscriptionType]) {
-      this.subscriptions[subscriptionType] = [];
-    }
     this.subscriptions[subscriptionType].push(subscriptionId);
     return subscriptionId;
   }
 
   async unsubscribe(subscriptionId: string): Promise<boolean> {
-    Object.keys(this.subscriptions).map(e => {
-      this.subscriptions[e] = this.subscriptions[e]
-        ? this.subscriptions[e].filter(s => s !== subscriptionId)
-        : [];
+    Object.keys(this.subscriptions).forEach(method => {
+      this.subscriptions[method as Notification['method']] = this.subscriptions[
+        method as Notification['method']
+      ].filter((id: string) => id != subscriptionId);
     });
+
     return true;
   }
 
-  on(event: string, callback: EventEmitter.ListenerFn<any>): void {
-    this.events.on(event, callback);
-  }
+  on: OnType = (method, params) => this.events.on(method, params);
 
-  off(event: string, callback?: EventEmitter.ListenerFn<any> | undefined): void {
-    this.events.off(event, callback);
-  }
+  off: OffType = (method, params) => this.events.off(method, params);
 
   protected async onMessage(event: MessageEvent) {
     const message = event.data;
     if (!message.jsonrpc) {
       return;
     }
 
-    if (isJsonRpcNotification(message)) {
-      const eventName = message.method;
-      if (eventName === 'UIUpdate') {
+    if (isJsonRpcNotification<keyof NotificationType>(message)) {
+      // TODO: use schema validations as better type guards
+      const notificationMethod = message.method;
+      const notificationParams = message.params;
+      this.events.emit(notificationMethod, notificationParams);
```

Comment:
Here we strip out the JSON-RPC headers; the message just came across that interface so the rest of the app does not need this information. The implication is that the channel client never sees the `message`, only an event with event string `message.method` and data `message.params`.